### PR TITLE
Topic/serialize qp model cereal

### PR DESCRIPTION
**Pull request by fabinsch**

### CODE REVIEW

File: include/proxsuite/proxqp/dense/serialize.hpp [hunk 1]
```diff
@@ -0,0 +1,233 @@
...
+  Eigen::Matrix<_Scalar, _Rows, _Cols, _Options, _MaxRows, _MaxCols> const& m)
+{
+  int32_t rows = m.rows();
+  int32_t cols = m.cols();
+  ar(rows);
+  ar(cols);
+
+  for (int i = 0; i < rows; i++)
+    for (int j = 0; j < cols; j++)
+      ar(m(i, j));
```
Comment (jcarpent) [inline include/proxsuite/proxqp/dense/serialize.hpp]: Could you use here the fact thay you have a C table?

File: test/src/dense_qp_wrapper.cpp [hunk 2]
```diff
@@ -6798,4 +6799,37 @@ TEST_CASE("ProxQP::dense: init must be called before update")
...
+          qp_random.l,
+          qp_random.u);
+
+  proxsuite::serialization::saveToBinary(qp.model, "tmp/qp_model.bin");
+  proxsuite::serialization::loadFromBinary(qp.model, "tmp/qp_model.bin");
+
+  proxsuite::serialization::saveToJSON(qp.model, "tmp/qp_model.json");
+  proxsuite::serialization::loadFromJSON(qp.model, "tmp/qp_model.json");
+}
+#endif
```
Comment (jcarpent) [inline test/src/dense_qp_wrapper.cpp]: Could you add a dedicated file for testing the serialization only?

File: include/proxsuite/proxqp/dense/model.hpp [hunk 3]
```diff
@@ -118,6 +121,23 @@ struct Model
...
+            CEREAL_NVP(n_total),
+            CEREAL_NVP(H),
+            CEREAL_NVP(g),
+            CEREAL_NVP(A),
+            CEREAL_NVP(b),
+            CEREAL_NVP(C),
+            CEREAL_NVP(l),
+            CEREAL_NVP(u));
+  }
+#endif
```
Comment (jcarpent) [inline include/proxsuite/proxqp/dense/model.hpp]: You should add it to a dedicated file, as in Pinocchio and FCL

File: CMakeLists.txt [hunk 4]
```diff
@@ -62,6 +62,7 @@ option(BUILD_BINDINGS_WITH_AVX2_SUPPORT "Build the bindings with AVX2 support."
 option(BUILD_BINDINGS_WITH_AVX512_SUPPORT
        "Build the bindings with AVX512 support." ON)
 option(TEST_JULIA_INTERFACE "Run the julia examples as unittest" OFF)
+option(SERIALIZATION_QP_MODEL "Add method to serialize QP model" ON)
```
Comment (jcarpent) [inline CMakeLists.txt]: Not the right name

File: CMakeLists.txt [hunk 5]
```diff
@@ -136,6 +137,25 @@ if(BUILD_WITH_VECTORIZATION_SUPPORT)
...
+  endif()
+
+  target_include_directories(
+    proxsuite
+    INTERFACE
+      "$<BUILD_INTERFACE:${CMAKE_CURRENT_LIST_DIR}/external/cereal/include>"
+      "$<INSTALL_INTERFACE:external/cereal/include>")
+  target_compile_definitions(proxsuite INTERFACE PROXSUITE_WITH_SERIALIZATION)
+
+endif()
```
Comment (jcarpent) [inline CMakeLists.txt]: We should not install Cereal. We can check if it is present and download for Python bindings and tests.
We should not link the main exported target to cereal. 
Only if the user needs it, it should install it and linked to it.

**Review by jcarpent — CHANGES_REQUESTED** [review]

File: include/proxsuite/proxqp/dense/serialize.hpp (hunk 1 above)
Comment (jcarpent) [reply include/proxsuite/proxqp/dense/serialize.hpp]: it should be quicker.

File: CMakeLists.txt (hunk 5 above)
Comment (fabinsch) [reply CMakeLists.txt]: I do not really understand your comment. 
>Only if the user needs it, it should install it and linked to it.

This is what I am doing here I think. I set a CMake Option, that people can turn on if they want. If they turn this on, I make sure Cereal is present, I have to link the include directories etc...

File: include/proxsuite/proxqp/serialization/archive.hpp [hunk 6]
```diff
@@ -0,0 +1,231 @@
+//
```
Comment (jcarpent) [inline include/proxsuite/proxqp/serialization/archive.hpp]: To be moved to `include/proxsuite/serialization/`

File: include/proxsuite/proxqp/dense/wrapper.hpp [hunk 7]
```diff
@@ -11,6 +11,14 @@
...
 #include <proxsuite/proxqp/dense/preconditioner/ruiz.hpp>
 #include <chrono>
+// #ifdef PROXSUITE_WITH_SERIALIZATION
+// #include <fstream>
+// #include <string>
+// #include <proxsuite/helpers/serialize.hpp>
+// #include <proxsuite/helpers/filesystem.hpp>
+// #include <cereal/archives/binary.hpp>
+// #include <cereal/archives/json.hpp>
+// #endif
```
Comment (jcarpent) [inline include/proxsuite/proxqp/dense/wrapper.hpp]: to be removed

File: include/proxsuite/proxqp/dense/model.hpp [hunk 8]
```diff
@@ -9,6 +9,9 @@
 #include "proxsuite/linalg/veg/type_traits/core.hpp"
 #include "proxsuite/proxqp/dense/fwd.hpp"
 #include "proxsuite/proxqp/sparse/model.hpp"
+#ifdef PROXSUITE_WITH_SERIALIZATION
+#include <cereal/cereal.hpp>
+#endif
```
Comment (jcarpent) [inline include/proxsuite/proxqp/dense/model.hpp]: to be removed

File: bindings/python/src/expose-settings.hpp [hunk 9]
```diff
@@ -67,7 +73,23 @@ exposeSettings(pybind11::module_ m)
                    &Settings<T>::compute_preconditioner)
     .def_readwrite("update_preconditioner", &Settings<T>::update_preconditioner)
     .def_readwrite("verbose", &Settings<T>::verbose)
-    .def_readwrite("bcl_update", &Settings<T>::bcl_update);
+    .def_readwrite("bcl_update", &Settings<T>::bcl_update)
+    .def(pybind11::self == pybind11::self)
+    .def(pybind11::self != pybind11::self)
+#ifdef PROXSUITE_WITH_SERIALIZATION
```
Comment (jcarpent) [inline bindings/python/src/expose-settings.hpp]: to be removed

File: bindings/python/CMakeLists.txt [hunk 10]
```diff
@@ -49,8 +49,11 @@ function(CREATE_PYTHON_TARGET target_name COMPILE_OPTIONS dependencies)
   target_link_libraries(${target_name} PUBLIC ${dependencies} pybind11::module)
   target_compile_options(${target_name} PRIVATE ${COMPILE_OPTIONS})
   target_link_libraries(${target_name} PRIVATE proxsuite pybind11::module)
-  target_compile_definitions(${target_name}
-                             PRIVATE PYTHON_MODULE_NAME=${target_name})
+  target_compile_definitions(
+    ${target_name} PRIVATE PYTHON_MODULE_NAME=${target_name}
+                           PROXSUITE_WITH_SERIALIZATION)
```
Comment (jcarpent) [inline bindings/python/CMakeLists.txt]: by default, we should support serialization. So the WITH_SERIALIZATION var should be removed.

File: CMakeLists.txt [hunk 11]
```diff
@@ -136,15 +141,30 @@ if(BUILD_WITH_VECTORIZATION_SUPPORT)
...
-  LIBRARY DESTINATION ${CMAKE_INSTALL_FULL_LIBDIR}
-  ARCHIVE DESTINATION ${CMAKE_INSTALL_FULL_LIBDIR}
-  RUNTIME DESTINATION ${CMAKE_INSTALL_FULL_BINDIR})
+if(BUILD_TESTING OR BUILD_PYTHON_INTERFACE)
+  # Download cereal for pything bindings and unittests
+  set(cereal_dir ${PROJECT_SOURCE_DIR}/external/cereal)
+  set(cereal ${cereal_dir}/README.md)
+  find_package(Git REQUIRED)
+  if(NOT EXISTS ${cereal})
+    message("Initialize cereal submodule.")
```
Comment (jcarpent) [inline CMakeLists.txt]: the message should be removed.

File: include/proxsuite/proxqp/serialization/eigen.hpp [hunk 12]
```diff
@@ -0,0 +1,113 @@
+//
```
Comment (jcarpent) [inline include/proxsuite/proxqp/serialization/eigen.hpp]: To be moved to `include/proxsuite/serialization/`

File: test/src/serialization.hpp [hunk 13]
```diff
@@ -0,0 +1,102 @@
+#include <doctest.hpp>
```
Comment (jcarpent) [inline test/src/serialization.hpp]: Could you include the hpp file within the cpp file?

**Review by jcarpent — CHANGES_REQUESTED** [review]
I made some marginal comments which should enforce the genericity

**Pull request merged**